lib/index.js

```diff
@@ -0,0 +1,93 @@
+var Promise = require('bluebird');
+var Redis = require('redis-fast-driver');
+var transformations = require('./redis-result-transformation-functions');
+
+var juttle_utils = require('juttle/lib/runtime').utils;
+
+function RedisBackend(config, JuttleRuntime) {
+    var redis = Promise.promisifyAll(new Redis(config));
+    var connect_promise = new Promise(function(resolve, reject) {
+        redis.on('ready', resolve);
+        redis.on('error', reject);
+    });
+
+    var Read = JuttleRuntime.proc.base.extend({
+        procName: 'redis_read',
+        sourceType: 'batch',
+
+        initialize: function(options, params) {
+            if (!options.hasOwnProperty('raw')) {
+                throw new Error('must specify a redis query with -raw');
+            }
+
+            this.command = options.raw.split(' ');
+        },
+
+        start: function() {
+            var self = this;
+            var keyword = this.command[0].toLowerCase();
+
+            if (!transformations.hasOwnProperty(keyword)) {
+                return self.trigger('error', new Error('unsupported/invalid Redis command: ', keyword));
+            }
+            return connect_promise.then(function() {
+                return redis.rawCallAsync(self.command);
+            })
+            .then(function(result) {
+                var points = transformations[keyword](self.command, result);
+
+                self.emit(points);
+                self.emit_eof();
+            })
+            .catch(function(err) {
+                self.trigger('error', new Error(err.message));
+            });
+        }
+    });
+
+    var Write = JuttleRuntime.proc.sink.extend({
+        procName: 'redis_write',
+        initialize: function(options) {
+            var self = this;
+            this.eofs = 0;
+            this.in_progress_writes = 0;
+        },
+        process: function(points) {
+            var self = this;
+            this.in_progress_writes++;
+            connect_promise.then(function() {
+                return Promise.map(points, function(pt) {
+                    if (pt.hasOwnProperty('redis_write_command')) {
+                        var command = pt.redis_write_command.split(' ');
+                        return redis.rawCallAsync(command)
+                            .catch(function(err) {
+                                self.trigger('error', 'bad write command: ' + JSON.stringify(command) + ': ' + err.message);
+                            });
+                    }
+                });
+            })
+            .finally(function() {
+                self.in_progress_writes--;
+                self._maybe_done();
+            });
+        },
+        _maybe_done: function() {
+            if (this.eofs === this.ins.length && this.in_progress_writes === 0) {
```

Comment:
It doesn't! That was a lingering notion I had from the early days of Juttle that turns out not to be accurate anymore, since we now have smart merging that means the write will only get 1 EOF even if it has a parallel graph feeding into it. So this code is actually borken as it stands, so I'll get rid of the EOF-counting logic.